File: Week-02/Day-05/main.go
package main

import "fmt"

type Node struct {
	data int
	next *Node
	prev *Node
}

type CircularDoublyLinkedList struct {
	head *Node
	tail *Node
}

func main() {
	fmt.Println("Circular Doubly Linked Lists")

	list := &CircularDoublyLinkedList{}

	// InsertAtBeginning:
	list.InsertAtBeginning(71)
	list.InsertAtBeginning(22)

	fmt.Print("Printing List Forward: ")
	list.PrintListForward()

	// InsertAtEnd:
	list.InsertAtEnd(13)
	list.InsertAtEnd(81)

	fmt.Print("Printing List Backward: ")
	list.PrintListBackward()

	// DeleteNode:
	list.DeleteNode(81)
	fmt.Print("List after Deleting Node: ")
	list.PrintListForward()
}

// InsertAtBeginning inserts a new node at the beginning of the list
func (list *CircularDoublyLinkedList) InsertAtBeginning(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		newNode.next = newNode
		newNode.prev = newNode
		list.head = newNode
	} else {
		tail := list.head.prev
		newNode.next = list.head
		newNode.prev = tail
		list.head.prev = newNode
		tail.next = newNode
		list.head = newNode

	}
}

// InsertAtEnd inserts a new node at the end of the list
func (list *CircularDoublyLinkedList) InsertAtEnd(data int) {
	newNode := &Node{data: data}
	if list.head == nil {
		newNode.next = newNode
		newNode.prev = newNode
		list.head = newNode
		list.tail = newNode
	} else {
		tail := list.head.prev
		newNode.next = list.head // newNode's next should point to head
		newNode.prev = tail      // newNode's prev points to current tail
		tail.next = newNode
		list.head.prev = newNode // Update the tail to newNode
	}
}

// DeleteNode deletes a node with specific data from the list
func (list *CircularDoublyLinkedList) DeleteNode(data int) {
	if list.head == nil {
		fmt.Println("List is empty.")
		return
	}
	temp := list.head
	for temp.data != data {
		temp = temp.next
		if temp == list.head {
			fmt.Println("Node not found.")
			return
		}
	}

	if temp.next == temp { // list contains only one element

		list.head = nil
	} else {
		temp.prev.next = temp.next
		temp.next.prev = temp.prev
		if temp == list.head {
			list.head = temp.next
		}
	}
}

// PrintListForward prints all the nodes in the list from head to tail
func (list *CircularDoublyLinkedList) PrintListForward() {
	if list.head == nil {
		fmt.Println("List is empty")
		return
	}
	temp := list.head
	for {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.next
		if temp == list.head {
			break
		}
	}
	fmt.Println(temp.data) // Print the data of the node pointing back to the head
}

// PrintListBackward prints all the nodes in the list from tail to head
func (list *CircularDoublyLinkedList) PrintListBackward() {
	if list.head == nil {
		fmt.Println("List is empty.")
		return
	}
	temp := list.head.prev

	for {
		fmt.Printf("%d -> ", temp.data)
		temp = temp.prev
		if temp.next == list.head.prev {
			break
		}
	}
	fmt.Println(temp.data) // Print the data of the node pointing back to the tail
}

File: Week-02/Day-05/main.cpp
#include <iostream>
#include <iterator>
using namespace std;

class CircularSinglyLinkedList {
private:
  struct Node {
    int data;
    Node* next;
    Node(int data): data(data), next(nullptr) {}
  };
  Node* head;
public:
  CircularSinglyLinkedList(): head(nullptr){}
  ~CircularSinglyLinkedList();

  void insertAtBeginning(int data);
  void insertAtEnd(int data);
  void insertAtPosition(int data, int position);
  void printList();
  void deleteNode(int key);
};

CircularSinglyLinkedList::~CircularSinglyLinkedList(){
if(!head) return;
  Node* current = head;
  Node* nextNode;
  do{
    nextNode = current->next;
    delete current;
    current = nextNode;
  } while (current != head);
}

// Doubly linked list 

class CircularDoublyLinkedList{
private:
  struct Node {
    int data;
    Node* next;
    Node* prev;
    Node(int data): data(data), next(nullptr), prev(nullptr) {}
  };
    Node* head;

public:
  CircularDoublyLinkedList(): head(nullptr) {}
  ~CircularDoublyLinkedList();

  void insertAtBeginning(int data);
  void insertAtEnd(int data);
  void insertAtPosition(int position, int data);
  void deleteNode(int key);
  void printList();
  
};

CircularDoublyLinkedList::~CircularDoublyLinkedList() {
  if(!head) return;
  Node* current = head;
  do{
    Node* nextNode = current->next;
    delete current;
    current = nextNode;
  } while(current != head);
}

void CircularSinglyLinkedList::insertAtBeginning(int data){
  Node* newNode = new Node(data);
  if(!head){
    head = newNode;
    newNode->next = head;
  }else {
    Node* temp = head;
    while (temp->next != head) {
      temp = temp->next;
    }
    newNode->next = head;
    temp->next = newNode;
    head = newNode;
  }
}

void CircularSinglyLinkedList::insertAtEnd(int data){
  Node* newNode = new Node(data);
  if (!head) {
    head = newNode;
    newNode->next = head;
  } else {
    Node* temp = head;
    while (temp->next != head) {
      temp = temp->next;
    }
    temp->next = newNode;
    newNode->next = head;
  }
}

// insertAtPosition
void CircularSinglyLinkedList::insertAtPosition(int data, int position){
  if(position == 0) {
    insertAtBeginning(data);
    return;
  }
  Node* newNode = new Node(data);
  Node* temp = head;
  for (int i = 0; i < position - 1 && temp->next != head; ++i){
    temp = temp->next;
  }
  if(temp->next == head){
    insertAtEnd(data);
  } else {
    newNode->next = temp->next;
    temp->next = newNode;
  }
}

// deleteNode
void CircularSinglyLinkedList::deleteNode(int key){
  if(!head) return;
  Node* current = head;  
  Node* prev = nullptr;
  do{
    if(current->data == key){
      if (prev) {
      prev->next = current->next;
      if(current == head) head = current->next;
    } else {
      Node* temp = head;
      while (temp->next != head) temp = temp->next;
      head = head->next;
      temp->next = head;
      }
    delete current;
    return;
    }
    prev = current;
    current = current->next;
    } while(current != head);

}

// printList
void CircularSinglyLinkedList::printList() {
  if (!head) return;
  Node* temp = head;
  do {
    cout << temp->data << " ";
    temp = temp->next;
  } while (temp != head);
  cout << endl;
}


// InsertAtBeginning
void CircularDoublyLinkedList:: insertAtBeginning(int data){
    Node* newNode = new Node(data);
    if(!head){
      head = newNode;
      newNode->next = head;
      newNode->prev = head;
    } else {
      Node* temp = head;
      while (temp->next != head) {
        temp = temp->next;
      }
      newNode->next = head;
      newNode->prev = temp;
      head->prev = newNode;
      temp->next = newNode;
      head = newNode;
    }
}

// InsertAtEnd
void CircularDoublyLinkedList:: insertAtEnd(int data){
  Node* newNode = new Node(data);
    if(!head) {
      head = newNode;
      newNode->next = head;
      newNode->prev = head;
    } else {
      Node* temp = head;
      while (temp->next != head) {
        temp = temp->next;
      }
      temp->next = newNode;
      newNode->next = head;
      newNode->prev = temp;
      head->prev = newNode;
    }
}

// insertAtPosition
void CircularDoublyLinkedList::insertAtPosition(int data, int position){
  if (position == 0){
    insertAtBeginning(data);
    return;
  } 

  Node* newNode = new Node(data);
  Node* temp = head;
  for(int i = 0; i < position-1 && temp->next != head; ++i){
    temp = temp->next;
  }

  if(temp->next == head){
    insertAtEnd(data);
  } else {
    newNode->next = temp->next;
    temp->next->prev = newNode;
    temp->next = newNode;
    newNode->prev = temp;
  }
}

//deleteNode
void CircularDoublyLinkedList::deleteNode(int key){
  if(!head) return;
  Node* current = head;
  do{
    if(current->data == key) {
      if(current->prev){
        current->prev->next = current->next;
        current->next->prev = current->prev;
      } else {
        Node* temp = head;
        while (temp->next != head) temp = temp->next;
        head = head->next;
        head->prev = temp;
        temp->next = head;
      }
      delete current;
      return;
    }
    current = current->next;
  } while (current != head);
}


void CircularDoublyLinkedList::printList() {
  if (!head) return;
  Node* temp = head;
  do {
    cout << temp->data << " ";
    temp = temp->next;
  } while (temp != head);
  cout << endl;
}

int main () {
  // circularSinglyList
  cout << " Circular Singly Linked List: \n";
  CircularSinglyLinkedList listS;
  listS.insertAtEnd(21);
  listS.insertAtBeginning(71);
  listS.insertAtPosition(2,2);
  listS.printList();
  listS.deleteNode(2);
  listS.printList();

  // Doubly
  cout << " Circular Doubly Linked List: \n";
  CircularDoublyLinkedList listD;
  listD.insertAtBeginning(54);
  listD.insertAtBeginning(18);
  listD.insertAtPosition(2,5);
  listD.insertAtBeginning(23);
  listD.insertAtEnd(7);
  listD.printList();
  listD.deleteNode(2);
  listD.printList();



  return 0;
}











